### PR TITLE
SC: Add ability to cancel update initiated by self

### PR DESCRIPTION
**Pull request by velzevur**
PT [Add ability to cancel update initiated by self() ](https://www.pivotaltracker.com/story/show/168289008)

Protocol update: https://github.com/aeternity/protocol/pull/417

### CODE REVIEW

File: apps/aechannel/src/aesc_fsm.erl [hunk 1]
```diff
@@ -3974,6 +3985,39 @@ handle_call(St, Req, From, #data{} = D) ->
...
+    case {lists:member(Tag, CancelMyTags), lists:member(Tag, CancelAckTags)} of
+        {true, _} ->
+            report(info, canceled_update, D),
+            lager:debug("update canceled", []),
+            next_state(open, clear_ongoing(D#data{op = ?NO_OP}),
+                      [{reply, From, ok}]);
+        {_, true} ->
+            report(info, canceled_update, D),
+            lager:debug("update canceled", []),
+            handle_update_conflict(Tag, D, [{reply, From, ok}]);
```
Comment (uwiger) [inline apps/aechannel/src/aesc_fsm.erl]: Shouldn't this rather be:
```erlang
handle_recoverable_error(#{ code => ?ERR_ABORT
                          , respond => true
                          , msg_type => Tag }, D, [{reply, From, ok}]);
```

File: apps/aehttp/test/aehttp_sc_SUITE.erl [hunk 2]
```diff
@@ -3981,3 +4006,259 @@ sc_ws_pinned_contract(Cfg) ->
...
+
+            %% on the other hand, responder can cancel it
+            {ok, #{<<"event">> := Event}} = wait_for_channel_event(RConnPid, info, Cfg),
+            {ok, _Tag, _Data} = wait_for_channel_event(RConnPid, sign, Cfg),
+            ok = ?WS:unregister_test_for_channel_events(RConnPid, [info, sign]),
+
+            ok = ?WS:register_test_for_channel_events(IConnPid, [conflict]),
+            sc_ws_cancel_update_(RConnPid, Cfg),
+            {ok, #{ <<"error_code">> := 2
+                  , <<"error_msg">> := <<"conflict">> }} =
```
Comment (uwiger) [inline apps/aehttp/test/aehttp_sc_SUITE.erl]: If the proposed change above is made, this needs to change to:
```
, <<"error_msg">> := <<"abort">> }} =
```

**Comment by velzevur** [comment]
Found a bug. Log:
```

*** User 2019-09-27 13:40:39.658 ***
[responder] Sending to server <<"{\"jsonrpc\":\"2.0\",\"method\":\"channels.cancel\",\"params\":{}}">>


*** User 2019-09-27 13:40:39.667 ***
[responder] Received msg #{<<"jsonrpc">> => <<"2.0">>,
                           <<"method">> => <<"channels.info">>,
                           <<"params">> =>
                               #{<<"channel_id">> =>
                                     <<"ch_95YaTDZAysRu3GkmW2yKkCK1H4fGtcttoj2qwFDfUSduTpCPf">>,
                                 <<"data">> =>
                                     #{<<"event">> => <<"canceled_update">>}},
                           <<"version">> => 1}


*** User 2019-09-27 13:40:39.667 ***
Origin = channel, Action = info, Tag = undefined


*** User 2019-09-27 13:40:39.667 ***
Register = {register,#{<0.5335.0> => #Ref<0.1508090406.1120927746.5882>},
                     #{{channel,error} => [<0.5335.0>],
                       {channel,info} => [<0.5335.0>]}}


*** User 2019-09-27 13:40:39.668 ***
[initiator] Received msg #{<<"jsonrpc">> => <<"2.0">>,
                           <<"method">> => <<"channels.conflict">>,
                           <<"params">> =>
                               #{<<"channel_id">> =>
                                     <<"ch_95YaTDZAysRu3GkmW2yKkCK1H4fGtcttoj2qwFDfUSduTpCPf">>,
                                 <<"data">> =>
                                     #{<<"channel_id">> =>
                                           <<"ch_95YaTDZAysRu3GkmW2yKkCK1H4fGtcttoj2qwFDfUSduTpCPf">>,
                                       <<"error_code">> => 4,
                                       <<"error_msg">> => <<"abort">>,
                                       <<"round">> => 1}},
                           <<"version">> => 1}


*** User 2019-09-27 13:40:39.668 ***
Origin = channel, Action = conflict, Tag = undefined


*** User 2019-09-27 13:40:39.668 ***
Register = {register,#{<0.5335.0> => #Ref<0.1508090406.1120927746.5880>},
                     #{{channel,conflict} => [<0.5335.0>]}}


*** User 2019-09-27 13:40:39.668 ***
[responder] Received msg #{<<"jsonrpc">> => <<"2.0">>,
                           <<"method">> => <<"channels.conflict">>,
                           <<"params">> =>
                               #{<<"channel_id">> =>
                                     <<"ch_95YaTDZAysRu3GkmW2yKkCK1H4fGtcttoj2qwFDfUSduTpCPf">>,
                                 <<"data">> =>
                                     #{<<"channel_id">> =>
                                           <<"ch_95YaTDZAysRu3GkmW2yKkCK1H4fGtcttoj2qwFDfUSduTpCPf">>,
                                       <<"error_code">> => 4,
                                       <<"error_msg">> => <<"abort">>,
                                       <<"round">> => 1}},
                           <<"version">> => 1}

```

Flow is:
1. responder cancels an update
2. responder receives an info with `canceled_update` that the update had been canceled
3. initiator receives a `channels.conflict` with `error_msg` being `abort`. Now initiator knows that the ongoing update is canceled by the responder
4. Initiator sends a conflict back to the responder with the `abort`

Do we need step 4?

**Comment by AleksandarFilipov** [comment]
> Flow is:
> 
> 1. responder cancels an update
> 2. responder receives an info with `canceled_update` that the update had been canceled
> 3. initiator receives a `channels.conflict` with `error_msg` being `abort`. Now initiator knows that the ongoing update is canceled by the responder
> 4. Initiator sends a conflict back to the responder with the `abort`
> 
> Do we need step 4?

Nope, step 3 clearly states to the initiator that the update has been aborted on purpose and responder clearly know that it cancelled the update.

**Comment by velzevur** [comment]
Yes, fixed. Step 4 is removed

**Comment by velzevur** [comment]
WIP till #2801 is merged

File: apps/aechannel/src/aesc_fsm.erl [hunk 3]
```diff
@@ -2809,31 +2811,57 @@ send_update_ack_msg(SignedTx, #data{ on_chain_id = OnChainId
...
     maybe_send_error_msg(Info, Sn, Msg, Data),
-    report(conflict, Msg, Data),
+    case ReportConflict of
+        true -> report(conflict, Msg, Data);
+        false -> pass
+    end,
     log(snd, error_msg_type(Info, Data), Msg, Data).
 
+
+
```
Comment (uwiger) [inline apps/aechannel/src/aesc_fsm.erl]:
```suggestion
```

File: apps/aehttp/test/aehttp_sc_SUITE.erl [hunk 4]
```diff
@@ -1565,8 +1590,8 @@ sc_ws_contract_generic_(Origin, ContractSource, Fun, Config, Opts) ->
     [Fun(Owner, GetVolley, CreateContract, SenderConnPid,
          AckConnPid, OwnerPubkey, OtherPubkey, Opts, Config)
         || {Owner, {OwnerPubkey, OtherPubkey}} <- Actors],
-    ok = ?WS:unregister_test_for_channel_events(SenderConnPid, [sign, info, get, error]),
-    ok = ?WS:unregister_test_for_channel_events(AckConnPid, [sign, info, get, error]),
+    ok = ?WS:unregister_test_for_channel_events(SenderConnPid, EventTags),
```
Comment (uwiger) [inline apps/aehttp/test/aehttp_sc_SUITE.erl]: The function `with_registered_events(EventTags, [SenderConnPid, AckConnPid], fun() -> ... end)` could be used here.

File: apps/aehttp/test/aehttp_sc_SUITE.erl [hunk 5]
```diff
@@ -4208,3 +4239,259 @@ wait_for_expected_events(Pid, Config, ExpectedEvents) ->
...
+        end,
+    cancel_update(InitUpdate, Cfg, can_cancel_ack),
+    sc_ws_close_(Cfg),
+    ok.
+
+sc_ws_cancel_deposit(Cfg0) ->
+    Cfg = sc_ws_open_(Cfg0),
+    InitUpdate =
+        fun(WhoPid) ->
+            ok = ?WS:register_test_for_channel_events(WhoPid, [sign]),
```
Comment (uwiger) [inline apps/aehttp/test/aehttp_sc_SUITE.erl]: The function `with_registered_events([sign], [WhoPid], fun() -> ... end)` could be used here.

Comment (uwiger) [reply apps/aehttp/test/aehttp_sc_SUITE.erl]: ... and in similar patterns below

Comment (velzevur) [reply apps/aehttp/test/aehttp_sc_SUITE.erl]: Addressed

File: apps/aechannel/src/aesc_fsm.erl [hunk 6]
```diff
@@ -4160,6 +4186,42 @@ handle_call(St, Req, From, #data{} = D) ->
 handle_call(_St, _Req, From, D) ->
     keep_state(D, [{reply, From, {error, unknown_request}}]).
 
+handle_call_(awaiting_signature, cancel_update, From, #data{} = D) ->
+    #data{op = #op_sign{tag = Tag}} = D,
+    %% cancelable tags for an action initiated by me:
+    CancelMyTags = [ slash_tx
```
Comment (tolbrino) [inline apps/aechannel/src/aesc_fsm.erl]: These lists should be put into the `.hrl` file.

File: apps/aechannel/src/aesc_fsm.erl [hunk 7]
```diff
@@ -4160,6 +4186,42 @@ handle_call(St, Req, From, #data{} = D) ->
...
+                                      , respond => true
+                                      , msg_type => Tag }, D, [{reply, From, ok}],
+                                     _ReportConflictToClient = false);
+        {false, false} ->
+            lager:debug("update can not be canceled for ~p", [Tag]),
+            keep_state(D, [{reply, From, {error, not_allowed_now}}])
+    end;
+handle_call_(_NonSigningState, cancel_update, From, #data{} = D) ->
+    lager:debug("update can not be canceled while ~p", [_NonSigningState]),
+    keep_state(D, [{reply, From, {error, not_allowed_now}}]);
```
Comment (tolbrino) [inline apps/aechannel/src/aesc_fsm.erl]: The error tag is somewhat weird. How about `invalid_state`?

Comment (velzevur) [reply apps/aechannel/src/aesc_fsm.erl]: Well the state is actually valid, it is that the action is not allowed at the current state. The actual message propagated to the client is `Not allowed at current channel state` as described [here](https://github.com/aeternity/aeternity/blob/0d73c39ed467bb68b62a1b42af57f97d732addba/apps/aehttp/src/sc_ws_api_jsonrpc.erl#L262)

File: docs/release-notes/next/PT-168289008_SC_soft_reject.md [hunk 8]
```diff
@@ -0,0 +1 @@
+* Adds a functionality to reject updates in State Channel environment.
```
Comment (tolbrino) [inline docs/release-notes/next/PT-168289008_SC_soft_reject.md]:
```suggestion
* State Channels: Adds the ability to have a client cancel updates which the client itself has initiated.
```

**Comment by tolbrino** [comment]
@velzevur This needs a rebase now, sorry about that.

**Comment by velzevur** [comment]
`protocol` PR is comming up

**Pull request merged**